plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/MigrateProcedure.java

```diff
@@ -0,0 +1,354 @@
+        return new Schema(icebergSchema.asStructType().fields());
+    }
+
+    public Map<String, Optional<Partition>> listAllPartitions(HiveMetastore metastore, io.trino.plugin.hive.metastore.Table table)
+    {
+        List<String> partitionNames = table.getPartitionColumns().stream().map(Column::getName).collect(toImmutableList());
+        Optional<List<String>> partitions = metastore.getPartitionNamesByFilter(table.getDatabaseName(), table.getTableName(), partitionNames, TupleDomain.all());
+        if (partitions.isEmpty()) {
+            return ImmutableMap.of();
+        }
+        return metastore.getPartitionsByNames(table, partitions.get());
+    }
+
+    private List<DataFile> buildDataFiles(ConnectorSession session, HiveStorageFormat format, String location, PartitionSpec partitionSpec, StructLike partition, NameMapping nameMapping)
+            throws IOException
+    {
+        // TODO: Introduce parallelism
+        TrinoFileSystem fileSystem = fileSystemFactory.create(session);
+        FileIterator files = fileSystem.listFiles(location);
+        ImmutableList.Builder<DataFile> dataFiles = ImmutableList.builder();
+        while (files.hasNext()) {
+            FileEntry file = files.next();
+            if (!file.path().contains("/_") && !file.path().contains("/.")) {
+                Metrics metrics = loadMetrics(fileSystem, format, file.path(), nameMapping);
+                DataFile dataFile = buildDataFile(file, partition, partitionSpec, format.name(), metrics);
+                dataFiles.add(dataFile);
+            }
+        }
+        return dataFiles.build();
+    }
+
+    private Metrics loadMetrics(TrinoFileSystem fileSystem, HiveStorageFormat storageFormat, String path, NameMapping nameMapping)
+    {
+        InputFile inputFile = new ForwardingInputFile(fileSystem.newInputFile(path));
+        return switch (storageFormat) {
+            case ORC -> OrcMetrics.fromInputFile(inputFile, METRICS_CONFIG, nameMapping);
+            case PARQUET -> ParquetUtil.fileMetrics(inputFile, METRICS_CONFIG, nameMapping);
+            case AVRO -> new Metrics(Avro.rowCount(inputFile), null, null, null, null);
+            default -> throw new TrinoException(NOT_SUPPORTED, "Unsupported storage format: " + storageFormat);
+        };
+    }
+
+    private static List<String> toPartitionFields(io.trino.plugin.hive.metastore.Table table)
+    {
+        ImmutableList.Builder<String> fields = ImmutableList.builder();
+        fields.addAll(getPartitionColumnNames(table));
+        table.getStorage().getBucketProperty()
+                .ifPresent(bucket -> {
+                    if (bucket.getBucketedBy().size() > 1) {
+                        throw new TrinoException(NOT_SUPPORTED, "Cannot migrate table bucketed on multiple columns: " + bucket.getBucketedBy());
+                    }
+                    fields.add("bucket(%s, %s)".formatted(getOnlyElement(bucket.getBucketedBy()), bucket.getBucketCount()));
```

Comment:
I don't think we can convert bucketed tables, since Iceberg and Hive use different bucketing functions, which means the values will be in the wrong files.

Reply:
Also, Iceberg only supports a subset of types for bucketing.

Reply:
Thanks, let me remove support for bucketed tables.